lib/frrscript.c

```diff
@@ -102,67 +102,126 @@ static void codec_free(struct codec *c)
+		XCALLOC(MTYPE_SCRIPT, sizeof(struct lua_function_state));
+	lfs->name = tmp->name;
+	lfs->L = tmp->L;
+	return lfs;
+}
+
+static void lua_function_free(struct hash_bucket *b, void *data)
+{
+	struct lua_function_state *lfs = (struct lua_function_state *)b->data;
+	lua_close(lfs->L);
+	XFREE(MTYPE_SCRIPT, lfs);
+}
+
+/* internal frrscript APIs */
+
+int _frrscript_call_lua(struct lua_function_state *lfs, int nargs)
+{
+
+	int ret;
+	ret = lua_pcall(lfs->L, nargs, 1, 0);
 
 	switch (ret) {
 	case LUA_OK:
 		break;
 	case LUA_ERRRUN:
-		zlog_err("Script '%s' runtime error: %s", fs->name,
-			 lua_tostring(fs->L, -1));
+		zlog_err("Lua hook call '%s' : runtime error: %s", lfs->name,
+			 lua_tostring(lfs->L, -1));
 		break;
 	case LUA_ERRMEM:
-		zlog_err("Script '%s' memory error: %s", fs->name,
-			 lua_tostring(fs->L, -1));
+		zlog_err("Lua hook call '%s' : memory error: %s", lfs->name,
+			 lua_tostring(lfs->L, -1));
 		break;
 	case LUA_ERRERR:
-		zlog_err("Script '%s' error handler error: %s", fs->name,
-			 lua_tostring(fs->L, -1));
+		zlog_err("Lua hook call '%s' : error handler error: %s",
+			 lfs->name, lua_tostring(lfs->L, -1));
 		break;
 	case LUA_ERRGCMM:
-		zlog_err("Script '%s' garbage collector error: %s", fs->name,
-			 lua_tostring(fs->L, -1));
+		zlog_err("Lua hook call '%s' : garbage collector error: %s",
+			 lfs->name, lua_tostring(lfs->L, -1));
 		break;
 	default:
-		zlog_err("Script '%s' unknown error: %s", fs->name,
-			 lua_tostring(fs->L, -1));
+		zlog_err("Lua hook call '%s' : unknown error: %s", lfs->name,
+			 lua_tostring(lfs->L, -1));
 		break;
 	}
 
 	if (ret != LUA_OK) {
-		lua_pop(fs->L, 1);
+		lua_pop(lfs->L, 1);
 		goto done;
 	}
 
+	if (lua_gettop(lfs->L) != 1) {
+		zlog_err(
+			"Lua hook call '%s': Lua function should return only 1 result",
+			lfs->name);
+		ret = 1;
+		goto done;
+	}
+
+	if (lua_istable(lfs->L, 1) != 1) {
+		zlog_err(
+			"Lua hook call '%s': Lua function should return a Lua table",
+			lfs->name);
+		ret = 1;
+	}
+
 done:
 	/* LUA_OK is 0, so we can just return lua_pcall's result directly */
 	return ret;
 }
 
-void *frrscript_get_result(struct frrscript *fs,
-			   const struct frrscript_env *result)
+void *frrscript_get_result(struct frrscript *fs, const char *function_name,
+			   const char *name,
+			   void *(*lua_to)(lua_State *L, int idx))
 {
-	void *r;
-	struct frrscript_codec c = {.typename = result->typename};
+	void *p;
+	struct lua_function_state *lfs;
+	struct lua_function_state lookup = {.name = function_name};
 
-	struct frrscript_codec *codec = hash_lookup(codec_hash, &c);
-	assert(codec && "No encoder for type");
+	lfs = hash_lookup(fs->lua_function_hash, &lookup);
 
-	if (!codec->decoder) {
-		zlog_err("No script decoder for type '%s'", result->typename);
+	if (lfs == NULL)
 		return NULL;
-	}
 
-	lua_getglobal(fs->L, result->name);
-	r = codec->decoder(fs->L, -1);
-	lua_pop(fs->L, 1);
+	/* results table is idx 1 on the stack, getfield pushes our item to idx
+	 * 2
```

Comment:
1. The results table is on the top of the stack right?
2. getfield pushes your item to the top of the stack

right? IMO better to operate with negative indices from the top of the stack than hardcode positions.

Reply:
updated in https://github.com/FRRouting/frr/pull/8982/commits/7300c0194493c261fbf54f4b2f4eb539fdea217b. Also added asserts to check if the returned Lua table is the only thing on the stack, before and after the lua_to decoder runs.